go/vt/topo/keyspace_routing_rules_lock.go

```diff
@@ -0,0 +1,66 @@
+/*
+Copyright 2024 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package topo
+
+import (
+	"context"
+	"fmt"
+	"path"
+
+	"vitess.io/vitess/go/vt/log"
+)
+
+// KeyspaceRoutingRulesLock is a wrapper over TopoLock, to serialize updates to the keyspace routing rules.
+type KeyspaceRoutingRulesLock struct {
+	*TopoLock
+}
+
+func createTopoDirIfNeeded(ctx context.Context, ts *Server) error {
```

Comment:
IMO, you should not be able to lock something that doesn't exist. Creating it just to lock it seems wrong and it's not how we handle topo locks generally AFAIK. If this is simply about the prefix/lock path itself, that is handled by the topo server when using `ts.globalCell.Lock()`. Why are we not using that here?

Reply:
Also please note that I don't believe the method used here will not work for ZooKeeper. Each topo server has special handling of lock files. In the case of ZK2 it's prefix/lock{0,1,...} and it's only if you were able to create lock0 that you have acquired the lock. All of this is handled in the topo implementation itself. I think that we should be leveraging the proper layering here rather than trying to implement the lock handling at this upper layer.

Reply:
I think that I was wrong on AT LEAST some of this, so I'm going to resolve it and start over fresh with a full review and some local testing. 🙂 Sorry about that.